pyro/infer/abstract_infer.py

```diff
@@ -137,12 +67,43 @@ def _traces(self, *args, **kwargs):
         """
         raise NotImplementedError("inference algorithm must implement _traces")
 
+    def _marginal_single_site(self, site="_RETURN"):
+        sites_key = frozenset([site])
+        if sites_key in self._marginals:
+            return self._marginals[sites_key]
+        marginal = Empirical()
+        for tr, log_weight in zip(self.exec_traces, self.log_weights):
+            marginal.add(tr.nodes[site]["value"], log_weight=log_weight)
+        self._marginals[sites_key] = marginal
+        return marginal
+
+    def marginal(self, sites=None):
```

Comment:
Per design doc/discussion, what do you think of having this be a separate function/class instead of a method?

Reply:
There is a separate `Marginal` function, which is closer to the design doc. I think this relates to the caching issue that you mentioned above. 
 - The motivation is to populate traces once, and pass around this `TracePosterior` class to different methods, e.g. for visualization, summary stats etc. In that case, we will be doing repeated calls to `trace_posterior.marginal("site")` or other utility methods, which can be avoided. 
 - The above can be avoided by passing around these marginals to helper utilities, but that would get clunky with a lot of latent sites. What we can do is to memoize those calls, but it seems cleaner to have a functional, lazy `Marginal` and an imperative `.marginal` method.
 - The `Marginal` implementation above is lazy - it doesn't populate traces until we call `sample` or `log_prob`. While it memoizes the trace collection part, it is limited because the memoization is only based on `*args, **kwargs` ids, and if we use a different tensor instance for the same value, we will end up running the model and collecting traces again. To not have to run trace collection for each marginal `sample` or `log_prob` call (for non-Search algos), we will need a non-lazy version of Marginal which doesn't mesh well with Search.

This is a solution that I converged to while wrestling with some HMC stuff and our tests/examples (which we probably need more of), so I could easily be missing a lot of future use cases. Would love to hear suggestions from you and @fritzo to ensure that we have a generic reusable implementation, or at least one that can support those use cases in the future.

Reply:
As discussed in the design doc, I think the thing to give up on for now is the laziness in `Marginal`.  Instead we can bind a fixed set of input arguments to a posterior and have `Marginal` call `traces` in `Marginal.__init__`.  That way we can separate `marginal` and `TracePosterior` and directly pass around marginal distributions (or dictionaries of marginal distributions as mentioned below) to summarization helpers without having to maintain any extra mutable internal state.

Reply:
I think I will need your help to work it out. 

If we change to a non-lazy marginal like follows:
```
class EmpiricalMarginal(Distribution):
    """
    :param trace_dist: a TracePosterior instance representing a Monte Carlo posterior.

    Marginal distribution, that wraps over a TracePosterior object to provide a
    a Distribution over the return values of the TracePosterior's model. Calls
    ``TracePosterior.marginal()`` under the hood, but defers populating traces
    until ``.sample`` or ``.log_prob`` methods are called.
    """
    has_enumerate_support = True

    def __init__(self, trace_dist, *args, **kwargs):
        assert isinstance(trace_dist, TracePosterior), \
            "trace_dist must be trace posterior distribution object"

        self.sites = kwargs.pop("sites", "_RETURN")
        self.empirical_marginal = trace_dist(*args, **kwargs).marginal(self.sites)
        super(Distribution, self).__init__()

    def sample(self, sample_shape=torch.Size()):
        return self.empirical_marginal.sample(sample_shape)

    def log_prob(self, value):
        return self.empirical_marginal.log_prob(value)

    def enumerate_support(self):
        return self.empirical_marginal.enumerate_support()
```

and modify the schelling example, it throws an error about multiple sites named `loc`. I tried it earlier as well, and wasn't able to get it to work. Does it need changes in `Search.py`?

```
def alice(preference, depth):
    """
    Alice decides where to go by reasoning about Bob's choice
    """
    alice_prior = location(preference)
    return pyro.sample("bob_choice", EmpiricalMarginal(Search(bob), preference, depth - 1),
                       obs=alice_prior)


def bob(preference, depth):
    """
    Bob decides where to go by reasoning about Alice's choice
    """
    bob_prior = location(preference)
    if depth > 0:
        return pyro.sample("alice_choice", EmpiricalMarginal(Search(alice), preference, depth),
                           obs=bob_prior)
    else:
        return bob_prior
```

> without having to maintain any extra mutable internal state

Conspicuous mutability is the design choice of Python, unfortunately. Without immutable data structures and other supporting language features, many functional constructs in Python would be woefully inefficient or gnarly. :( 

Anyways, will sync with you in person, and try to work this out for Search.

Reply:
Maybe we should just delete `Search` and the Schelling examples, unless @ngoodman objects...

Reply:
Or we can have different versions of `Marginal` as you had in the design doc, and it can be lazy for Search? I think I was hoping that we can re-use the same, but it might just be easier to have different implementations instead.